Create Preview Loading screen

rt4914 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://developers.forem.com/contributing-guide/forem#create-a-pull-request
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
This PR introduces loading screen when user clicks on `Preview` while creating a post.

**Following features have been implemented**
1. When `Preview` is clicked it gets switched immediately.
2. Change title from `Create Post` to `Loading preview`
3. Disable the buttons in action bar at the bottom when preview is loading
4. The design of loading should match with [this](https://www.figma.com/file/PIPeiiVSADmcpCbiZTdJm9/Editor?node-id=137%3A0) where we have two designs -- with or without cover.

**How to test ?**
1. Sign-in into your local enviornment
2. Click on `Create Post`
3. Add some embed syntax , ideally for 2-3 different websites, so that it takes some time to load
4. Also test this by attaching a cover-image.


## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #17291 
- Closes #17291

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._



https://user-images.githubusercontent.com/9396084/175229933-989a0691-293f-4a8c-9825-ac345ecc6736.mov

<img width="400" alt="Screenshot 2022-06-23 at 11 48 53 AM" src="https://user-images.githubusercontent.com/9396084/175229970-db321c17-51f9-43dd-aaee-7b4f9a3f46bf.png">  <img width="400" alt="Screenshot 2022-06-23 at 11 49 04 AM" src="https://user-images.githubusercontent.com/9396084/175229971-9be8752b-4a3a-4d49-a26a-1a2670e7d72b.png">


### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

I have added test cases for this and there should not be any accessibility concerns.

**Question**: Where should [this](https://github.com/forem/forem/issues/17291#issuecomment-1109493910) change be applied? Should this be like this: `<span className="hidden s:inline-block mr-2 whitespace-nowrap aria-live">`
Or in a css file?

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [x] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

Not Sure

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

aitchiss commented on app/javascript/article-form/components/LoadingPreview.jsx:
```diff
@@ -0,0 +1,38 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+
+export const LoadingPreview = ({ version }) => {
```
I think feel free to add new CSS selectors if need be 👍 We have e.g. `crayons-story__body` for the article loading, but maybe we could create something like `crayons-preview__cover-image`, etc. It could go in the `article-form.scss`

rt4914 replied:
Based on your comment I have applied following changes:
1. In `acticle-form.scss` created `crayons-preview__cover` which uses `--body-bg` as background color.
2. For all other styles in LoadingPreview I have used the existing selectors like `crayons-scaffold`, `crayons-story__body`, `crayons-story__indention` ,etc.
3. I have add some new width values in `_generator.scss` which are useful in creating the UI.

I have done these changes as it was the easiest working solution for me but if there are any changes needed in these do let me know.

rt4914 commented:
@aitchiss This PR is almost finished with help needed in two things:
1. Where should [this](https://github.com/forem/forem/issues/17291#issuecomment-1109493910) change be applied? Should this be like this: `<span className="hidden s:inline-block mr-2 whitespace-nowrap aria-live">` i.e. adding `aria-live` to className
Or in a css file?
2. CI checks are not passing and I am unable to pinpoint which test is failing and why.

aitchiss commented:
@rt4914 `aria-live` is an attribute, so it could be something like this:

```html
<span aria-live="polite" className="screen-reader-only"></span>
```

Be sure to avoid using `hidden` className because this also hides the content from screen reader users (our intended audience!)

You'll need to dynamically update the text inside that `span` when the state changes to loading/loaded - this could be either with giving it a `ref` with `useRef` hook, or you could just add an ID to the `span` and get it with `getElementById`

I'll take a look at the cypress specs too 🤓

aitchiss commented:
Looks like we had an issue in `Preview.jsx` where we assumed `previewResponse` was always an object with `processed_html` key, but sometimes it's an empty string, which is causing the error. I recorded a quick Loom to explain - https://www.loom.com/share/3c3d6b3cbeae46be85ee1edce3c1e755

If we change the `useEffect` in `Preview.jsx` to the below, the tests start passing again:

```jsx
 useEffect(() => {
    if (previewResponse?.processed_html?.includes('twitter-timeline')) {
      attachTwitterTimelineScript();
    }
  }, [previewResponse]);
```

rt4914 commented:
> Looks like we had an issue in `Preview.jsx` where we assumed `previewResponse` was always an object with `processed_html` key, but sometimes it's an empty string, which is causing the error. I recorded a quick Loom to explain - https://www.loom.com/share/3c3d6b3cbeae46be85ee1edce3c1e755
> 
> If we change the `useEffect` in `Preview.jsx` to the below, the tests start passing again:
> 
> ```js
>  useEffect(() => {
>     if (previewResponse?.processed_html?.includes('twitter-timeline')) {
>       attachTwitterTimelineScript();
>     }
>   }, [previewResponse]);
> ```

Thanks a lot for this. Along with the suggested change, I have changed the empty string to empty object because it should be clear that its an object and not a string.

rt4914 commented:
> ```
> aria-live="polite"
> ```

I have tried applying this change but I have used only `aria-live="polite"` in `PageTitle.jsx`. I think it can be discussed in PR review.

rt4914 commented:
This particular test [should preview content of a post with a gist embed](https://github.com/forem/forem/blob/765ac08d18104e09b27665cf32aee2d0d4e5567d/cypress/integration/seededFlows/publishingFlows/previewPost.spec.js#L107) is failing in cypress tests. 

I understand that it is failing because I have introduced `LoadingPreview` and its directly checking for gist-preview. I know to fix this we have to check for LoadingPreview which can be done using `cy.findByTestId('loading-preview').should('be.visible');` and after which I added `cy.findByTestId('preview-response').should('be.visible');` to make sure that the preview is actually visible before checking the gist content. But at this line the test-case fails from which I am assuming that it does not wait for loading to finish first and then execute the second line. 

How can we fix this error?

Once this error gets fixed there will be other errors too in this file which will need to be fixed.

aitchiss commented:
> This particular test [should preview content of a post with a gist embed](https://github.com/forem/forem/blob/765ac08d18104e09b27665cf32aee2d0d4e5567d/cypress/integration/seededFlows/publishingFlows/previewPost.spec.js#L107) is failing in cypress tests.
> 
> I understand that it is failing because I have introduced `LoadingPreview` and its directly checking for gist-preview. I know to fix this we have to check for LoadingPreview which can be done using `cy.findByTestId('loading-preview').should('be.visible');` and after which I added `cy.findByTestId('preview-response').should('be.visible');` to make sure that the preview is actually visible before checking the gist content. But at this line the test-case fails from which I am assuming that it does not wait for loading to finish first and then execute the second line.
> 
> How can we fix this error?
> 
> Once this error gets fixed there will be other errors too in this file which will need to be fixed.

I had a look at this error, and the output notes it's not so much a problem with the spec, but an unhandled exception in our code:

```
 1) Post Editor
       v2 Editor
         should preview content of a post with a gist embed:
     TypeError: The following error originated from your application code, not from Cypress.
  > Cannot read properties of null (reading 'outerHTML')
When Cypress detects uncaught errors originating from your application it will automatically fail the current test.
This behavior is configurable, and you can choose to turn this off by listening to the `uncaught:exception` event.
https://on.cypress.io/uncaught-exception-from-application
      at _loop (webpack-internal:///./app/javascript/utilities/gist.js:78:39)
      at loadEmbeddedGists (webpack-internal:///./app/javascript/utilities/gist.js:87:7)
      at MutationObserver.callback (webpack-internal:///./app/javascript/utilities/gist.js:114:11)
```

I took a look at [the offending line](https://github.com/forem/forem/blob/def5de6d28b0621d51b669fd71425254aa5ab29f/app/javascript/utilities/gist.js#L34), where we access `outerHTML` on an element that's apparently `null`...

[I did a wee loom to explain what's going on](https://www.loom.com/share/8d3fe41874ac40d3bce4127950e43456), but basically we want to change this clause in `articleForm.jsx`:

```js
 componentDidUpdate() {
    const { previewResponse } = this.state;

    if (previewResponse) {
      embedGists(this.formElement);
      this.constructor.handleRunkitPreview();
      this.constructor.handleAsciinemaPreview();
    }
  }
```

to this:

```jsx
 componentDidUpdate() {
    const { previewResponse } = this.state;

    if (previewResponse?.processed_html) {
      embedGists(this.formElement);
      this.constructor.handleRunkitPreview();
      this.constructor.handleAsciinemaPreview();
    }
  }
```

Because since we changed the `previewResponse` state to the proper object, we were meeting this `if` condition before we actually had the `processed_html`

rt4914 commented:
@Ridhwana @msarit This PR is ready for review now.

rt4914 commented:
**Update on the Travis CI failure**

rt4914 commented:
@Ridhwana @aitchiss This PR is ready for review.

aitchiss commented on app/javascript/article-form/components/LoadingPreview.jsx:
```diff
@@ -0,0 +1,32 @@
+};
+
+LoadingPreview.propTypes = {
+  version: PropTypes.string,
```
```suggestion
  version: PropTypes.oneOf(["default", "cover"]),
```
(nitpick) - we can help devs know how to implement this by adding the specific permitted props here 🙂

rt4914 replied:
This is really great. Did not knew about this.

aitchiss commented on app/javascript/article-form/components/PageTitle.jsx:
```diff
@@ -2,11 +2,19 @@ import { h } from 'preact';
     <div className="crayons-field__label flex items-center flex-1">
-      <span className="hidden s:inline-block mr-2 whitespace-nowrap">
-        Create Post
+      <span
+        aria-live="polite"
+        className="hidden s:inline-block mr-2 whitespace-nowrap"
+      >
+        {previewLoading ? 'Loading preview' : 'Create Post'}
```
I think we need to separate out the `aria-live`. It dynamically announces the changing text inside, so what we have at the moment is:

- "Create post"
- "Loading preview"
- "Create post"

when what we're trying to communicate to screen reader users is:

- Loading preview
- Preview loaded

I think the most straightforward way to do this would be to remove the `aria-live` from this `PageTitle` component (keeping the other changes so the visual title is correct), and instead add to `articleForem.jsx`:

```jsx
<span aria-live="polite" className="screen-reader-only">
  { previewLoading ? "Loading preview" : null }
  { previewShowing && !previewLoading ? "Preview loaded" : null }
</span>
```

Maybe there is a more elegant way to set the inner text content, but essentially I think we want:

- The inner text to be empty if we're currently on the "Edit" tab
- The inner text to be "Loading preview" when in loading state
- The inner text to be "Preview loaded" when on the preview tab but not in edit state

(the "screen-reader-only" class will make sure it's not visible to users, but it will still be announced by screen readers)

rt4914 replied:
I have applied this change and this is the output:

https://user-images.githubusercontent.com/9396084/177530102-e0b1a300-e9da-4d24-9a33-d5e40d30ae9f.mov

aitchiss commented on app/javascript/article-form/components/Preview.jsx:
```diff
@@ -80,17 +81,33 @@ const previewResponsePropTypes = PropTypes.shape({
 
+  if (previewLoading) {
+    const coverImage = articleState.mainImage;
+    let loadingPreview;
+    if (coverImage === null) {
+      loadingPreview = <LoadingPreview version="default" />;
+    } else {
+      loadingPreview = <LoadingPreview version="cover" />;
+    }
```
```suggestion
    const loadingPreview =  <LoadingPreview version={coverImage === null ? "default" : "cover"} />
```
just a suggestion / not blocking - it could even be in-lined on line 106

rt4914 replied:
Applied this change.

aitchiss commented on app/javascript/article-form/components/__tests__/LoadingPreview.test.jsx:
```diff
@@ -0,0 +1,19 @@
+    expect(results).toHaveNoViolations();
+  });
+
+  it('should render', () => {
```
(nitpick / not blocking) - maybe we should check it renders with the "cover" variant too?

rt4914 replied:
Added test for `should render with cover image`

aitchiss commented on app/javascript/article-form/components/__tests__/Options.test.jsx:
```diff
@@ -50,13 +50,66 @@ describe('<Options />', () => {
+    expect(getByTitle('Post options')).toBeDisabled();
+  });
+
+  it('shows the button is disabled when preview is not loading', () => {
```
```suggestion
  it('shows the button is not disabled when preview is not loading', () => {
```

😄

aitchiss commented on app/javascript/article-form/components/__tests__/Preview.test.jsx:
```diff
@@ -67,9 +67,13 @@ describe('<Preview />', () => {
   });
 
-  it('should have no a11y violations', async () => {
+  // no cover for loading
+  // with cover for loading
+
```
```suggestion
```

Maybe just left over by mistake?

aitchiss reviewed (approved):
Nice one! Thanks for making those changes - looks and works great ✨

rt4914 commented:
@aitchiss This PR is ready for another pass. I have applied all the changes that you suggested. Thanks.

rt4914 commented:
@Ridhwana This PR is ready for review. Thanks.
